credoai/modules/model_modules/privacy.py

```diff
@@ -76,127 +124,195 @@ def run(self):
-                random_state=42,
-            )
+            ) = attack_assess[1]
 
+        ## Fit of attack model
+        if attack_details["fit"] == "train_test":
             # Split train and test further and fit the model
             attack.fit(x_train_attack, y_train_attack, x_test_attack, y_test_attack)
 
-        # Sets balancing -> This might become optional if we use other metrics, tbd
+        if attack_details["fit"] == "train_only":
+            attack.fit(x_train_assess)
+
+        ## Rebalancing of the assessment datasets
         x_train_bln, y_train_bln, x_test_bln, y_test_bln = self._balance_sets(
             x_train_assess, y_train_assess, x_test_assess, y_test_assess
         )
 
-        # Attack inference
-        train = attack.infer(x_train_bln, y_train_bln)
-        test = attack.infer(x_test_bln, y_test_bln)
+        ## Assessment
+        if attack_details["assess"] == "membership":
+            return self._assess_attack_membership(
+                attack, x_train_bln, y_train_bln, x_test_bln, y_test_bln
+            )
 
-        return self._assess_attack(train, test, accuracy_score)
+        if attack_details["assess"] == "attribute":
+            return self._assess_attack_attribute(attack, attack_details, x_test_bln)
 
-    def _predict_binary_class_matrix(self, x):
-        """`predict` that returns a binary class matrix
+    def _define_model_arguments(self, attack_details):
+        """
+        Collates the arguments to feed to the attack initialization.
 
+        Parameters
         ----------
-        x : features array
-            shape (nb_inputs, nb_features)
+        attack_details : dict
+            Dictionary containing all the attack details
 
         Returns
         -------
-        numpy.array
-            shape (nb_inputs, nb_classes)
+        dict
+            Named arguments dictionary for the attack function
         """
-        y = self.model.predict(x)
-        y_transformed = np.zeros((len(y), self.nb_classes))
-        for ai, bi in zip(y_transformed, y):
-            ai[bi] = 1
-        return y_transformed
+        arg_dict = {
+            "estimator": self.attacked_model,
+            "classifier": self.attacked_model,
+            "attack_feature": self.attack_feature,
+        }
+        return {i: arg_dict[i] for i in attack_details["attack"]["kwargs"]}
+
+    def _preprocess_data(self, *args) -> tuple:
+        """
+        Further split test and train dataset.
+
+        Parameters
+        ----------
+        args : dict
+            x_train, y_train, x_test, y_test. The order needs to be respected.
+
+        Returns
+        -------
+        tuple
+            Length 2 tuple, first elements contains the split of the train set,
+            the second element contains the split of the test set.
+        """
+
+        train_sets = train_test_split(
+            args[0], args[1], random_state=42, train_size=self.attack_train_ratio
```

Comment:
You are passing a list of `args`, which should be generic, but then expecting them to be of some specific form. These are no longer general args. These are more like `training_arrays` and `testing_arrays`.